Python 3 syntax compatibility

graeme-winter opened the pull request:
Following from discussions about Python 2 to 3 syntax changing we have made a large number of automated syntax changes for cctbx_project with a handful of manual fixes to ensure tests pass. Have also rebased recent changes from @phyy-nx over the top. 

We have deliberately not touched mmtbx as this is tricky for us to test (relies a lot on PHENIX regression data) and fable as not needed for dials. 

However we believe these changes move us towards P3 syntax compatibility - not true compatibility as we've not addressed e.g. bytes / strings / unicode here, but should help. 

Much kudos to @ndevenish and others at Diamond for the heavy lifting here. 

Should fix #80 though "async" keyword will still be an issue. 

Post-merge we are happy to help with adding tools to prevent syntax regression occurring, as we already have for xia2, dials and dxtbx.

graeme-winter commented:
N.B. this does _not_ make code which will run in a Python3 interpreter, only code which can be interpreted by one, so e.g. xrange etc. are not fixed. Those transformations can be applied in a subsequent pull request.

graeme-winter commented:
@bkpoon would like detailed description of what automated changes were applied here.

bkpoon commented:
Just which automatic fixes were applied. A list of names of the fixes from the conversion tool is fine.

graeme-winter commented:
futurize -f print_with_import -f exec -f except -f raise \
             -f numliterals -f tuple_params -f paren -wn .

graeme-winter commented:
Followed by some manual changes to mop up some of the more idiosyncratic motifs...

ndevenish commented:
Barring absolutely manual changes, the transformations run were:
1. A custom fixer was run to identify and fix the raise statements that were being invoked like functions (this is so rare apparently libfuturize had not encountered it before)
2. `futurize` was then run with the deliberate minimal transforms to make python 3 AST compatibility; the following transformations were run simultaneously:
    - `print_with_import` to fix the prints statements
    - `exec` to rewrite use of the `exec` statement
    - `except` to rewrite exception catching syntax
    - `raise` to rewrite most of the raise calls
    - `numliterals` was run to rewrite the one or two places that use octal literals
    - `tuple_params` was run to correct anywhere using tuple expansion in either a function call or a lambda; the automatic rename was not corrected
    - `paren` was run to add missed parentheses inside list comprehensions
3. A custom fixer was run to recombine the split `__future__` imports introduced by an incomplete `futurize` stage 2.

graeme-winter commented:
Since there was some criticism of this being incomplete as it was syntax-only am going a bit deeper with it - which will obviously increase the risk of conflicts but should also be more useful.

graeme-winter commented:
@bkpoon would be useful to know if anything in here breaks any of your local (i.e. non public) tests

Most likely would result from colliding variable names but would welcome feedback

graeme-winter commented:
Since keywords are clobbered all over the place https://github.com/cctbx/cctbx_project/commit/334ca558960bfa33130a65bca7c5d3bfd1cc149a

bkpoon commented:
I see that the tool was applied to `libtbx`. I left a few files untouched since it was unclear if people were still using the older way of building bundles, but that can be figured out later. Please undo the changes to `libtbx` that are unnecessary. Also, @nksauter modified `fable` to work, so revert any unnecessary changes for that directory. I think there was an extra `print` statement from debugging, so only that `print` statement should be fixed. The `boost/boost_adaptbx` directories are currently being modified, so please revert any changes to files that have already been modified.

Has anyone looked at the changes that the automated tool makes? It will make changes that are not necessary. That's why I usually apply one fix at a time to see the differences and only keep the sensible changes that lets the code run in Python 3. For example, in line 60 of `libtbx/tst_easy_mp.py`
**master**
https://github.com/cctbx/cctbx_project/blob/ae48defa0eeb5fc9c00c9105b79867d115956ffd/libtbx/tst_easy_mp.py#L60
**this branch**
https://github.com/cctbx/cctbx_project/blob/f572cf39f12473f187c64194a6ae307487f1af88/libtbx/tst_easy_mp.py#L60
I skipped the conversion of `range` into a `list` since it is not needed in Python 2 or 3, but this pull request added it back. The automated tool does other things like this (like changing a `dict` to a `list`), so the changes from the automated tool should be verified before being committed.

While the tests still pass, runtime is increased as a result of these extra changes. For the `scitbx` tests, it looks like the runtime has at least doubled. Here is the script I used for testing,
```bash
#!/bin/bash

nproc=72

for module in scitbx cctbx
do
  for i in `seq 1 10`
  do
    rm -fr * .comm .files.pkl .ligands.pml .status.pkl
    libtbx.run_tests_parallel module=${module} nproc=${nproc} >& /dev/null
    echo ${module} `grep "Tests finished" run_tests_parallel_zlog`
  done
done

for i in `seq 1 10`
do
  rm -fr * .comm .files.pkl .ligands.pml .status.pkl
  phenix_regression.test_all_parallel nproc=${nproc} >& /dev/null
  echo phenix_regression `grep "Tests finished" run_tests_parallel_zlog`
done
```

And here are some timing results in seconds for 10 trials. The "old time" is with the current `master` branch and the "new time" is using the `python-3-syntax-compatibility` branch.

| module | old time (mean/std dev) | new time (mean/std dev) |
|-|-|-|
| `scitbx` |16.3 (0.15) | 34.0 (0.38) |
| `cctbx` | 59.1 (0.71) | 64.0 (1.1) |
|` phenix_regression` | 368.4 (18.3)  | 401.2 (8.5) |

I do not have the breakdown of individual modules since most of the main modules are tested in `phenix_regression`, but I can generate them.

graeme-winter commented:
re: `list(range(things))` - the automated conversion is maintaining the existing behaviour and does not "know" how you will use it - true that manually inspecting where `list` has been added would be useful. 

I thought I had avoided touching fable as nothing in dials depends on it, will check. 

Runtime / scitbx - will look, suspect overly defensive conversions.

graeme-winter commented:
@bkpoon please can you point to changes to fable in here?

ndevenish commented:
> Also, @nksauter modified `fable` to work, so revert any unnecessary changes for that directory

Please can you point out where this changes fable - Graeme *explicitly* mentioned that we don't change this this in the second sentence:

>We have deliberately not touched ...... fable as not needed for dials.

My `git diff` doesn't show anything touching `fable`.? BTW you might want to take another look at fable. It looks like in your manual process, at least one file was missed and others are printing tuples everywhere because you forgot to import `print_functions`.

The boost_adaptbx changes seem entirely reasonable and look like they would merge cleanly to any extra work you've done. I'm happy to report that none of the changes in libtbx seem unnecessary.

>It will make changes that are not necessary. That's why I usually apply one fix at a time to see the differences and only keep the sensible changes that lets the code run in Python 3.

Yes, it will, and the list() is an obvious example of this, and can probably mostly be omitted. But it's semantically correct with what python 2 is doing and doesn't break anything . I'm not sure where your list(dict( is that you are complaining about but suspect that it's being used to implicitly iterate over keys.

Entirely my opinion: There are 12,800 changes in this extremely minimal commit. On a minute per change, this is 25 days of full time work. If you wanted to slowly do work that can be easily automated the time to start was years ago, and in any case it's probably quicker and *less* error-prone to undo changes you don't want than manually copy them over. re: Runtime; Again, speaking for myself - but I'm not sure it matters, especially about 30s/400s - time is not a correctness issue. There will extremely likely be other performance implications while converting, and it's pointless to make changes, let along run the test suite *ten times* at every step along the way. Profile and fix afterwards *in the correct interpreter* when we know the full differences.

Anthchirp commented:
Measuring the time tests take with libtbx.run_tests_parallel is not very meaningful as the full time can be dominated by test ordering especially when the variance between individual test times is large. A long running test that is started later than before will keep running after all other threads have finished. Running 10 instances does not account for this as it may be deterministic. You either need to compare individual test run times or the full runtime with nproc=1 or account for ordering effects by eg. integrating the CPU usage throughout.

In any case speed regressions should be addressed post-conversion.

Converting only some files is not helpful as this means the conversion is not complete and files in the library are deprecated without saying so. Standard technique would be to apply the  automatic conversion anyway (if nobody uses it then surely it does not matter, right?) and add a DeprecationWarning to top of those files.

Other comments have already been addressed so I will not repeat them here.

graeme-winter commented:
Re: unnecessary changes - `from builtins import range` what you are getting at? Even if this is not used it means that the code behaviour is the same with Python 2/3 so figured it was comparable to `from __future__ import division` etc. which have been mandatory a long while in cctbx.

`list(range(n))` - majority of these are in tests:

```
Silver-Surfer:cctbx_project graeme$ grep -R 'list(range(' .|grep tst|wc -l
     203
Silver-Surfer:cctbx_project graeme$ grep -R 'list(range(' .|grep -v tst|wc -l
     115
```

which could be contributing to the apparent change in run-time. However these could also be necessary if the tests expect something list-like to compare with a flex array...

graeme-winter commented:
TODO: automate transformation 

`flex.type(list(range(params)))` => `flex.type(range(params))`

graeme-winter commented:
> Re: unnecessary changes - `from builtins import range` what you are getting at? Even if this is not used it means that the code behaviour is the same with Python 2/3 so figured it was comparable to `from __future__ import division` etc. which have been mandatory a long while in cctbx.
> 
> `list(range(n))` - majority of these are in tests:
> 
> ```
> Silver-Surfer:cctbx_project graeme$ grep -R 'list(range(' .|grep tst|wc -l
>      203
> Silver-Surfer:cctbx_project graeme$ grep -R 'list(range(' .|grep -v tst|wc -l
>      115
> ```
> 
> which could be contributing to the apparent change in run-time. However these could also be necessary if the tests expect something list-like to compare with a flex array...

More of these are in tests than are apparent, as tests also buried in `__init__.py:exercise_1()` etc - https://github.com/cctbx/cctbx_project/blob/d83b77c69e0026b0abf7b7bf669e3f7435e3618b/scitbx/matrix/__init__.py#L1304 - makes it hard to spot that these are not running code.

Is putting test code in `__init__.py` a good idea? I thought the libtbx way was `tst_thing.py`

graeme-winter commented:
Times for `scitbx`

This branch on MacBook Pro with `nproc=6`

```
real	1m8.399s
user	5m4.041s
sys	0m32.044s
```

master:

```
real	0m38.127s
user	2m40.666s
sys	0m30.519s
```

Verified that something expensive is happening _in the tests_ - probably we are looking at `list(range())` casting unnecessarily. Will look. 

FWIW there are some tests which dominate run-time in master:

```
Distribution of test runtimes:
  0.0s - 2.3s: 116
  2.3s - 4.6s: 8
  4.6s - 6.9s: 4
  6.9s - 9.2s: 3
  9.2s - 11.5s: 2
  11.5s - 13.8s: 1
  13.8s - 16.1s: 0
  16.1s - 18.3s: 0
  18.3s - 20.6s: 0
  20.6s - 22.9s: 1
```

which as @Anthchirp suggested show up better when looking at total CPU time. Detail:

```
Warning: the following are 5 longest jobs:
  libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/lbfgs/tst_lbfgs_fem.py": 22.9s
  libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/rigid_body/essence/tst_tardy.py": 13.0s
  libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tests/tst_gaussian.py": 10.3s
  libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tests/tst_r3_rotation.py": 9.3s
  libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tst_zernike.py": 9.2s
Please try to reduce overall runtime - consider splitting up these tests.
```

graeme-winter commented:
Though interestingly when run alone this test

```
libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/lbfgs/tst_lbfgs_fem.py"
```

takes wall clock 15.98 seconds on master and 16.12 on the branch, so a lot quicker, while

```
 libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tests/tst_curve_fitting.py"
```

takes a highly random amount of time - on master / this / this / master below:

```
Silver-Surfer:tmp graeme$ time libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tests/tst_curve_fitting.py"
OK

real	0m41.428s
user	0m41.180s
sys	0m0.199s
Silver-Surfer:tmp graeme$ time libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tests/tst_curve_fitting.py"
OK

real	0m3.942s
user	0m3.809s
sys	0m0.125s
Silver-Surfer:tmp graeme$ time libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tests/tst_curve_fitting.py"
OK

real	0m3.798s
user	0m3.675s
sys	0m0.116s
Silver-Surfer:tmp graeme$ time libtbx.python "/Users/graeme/svn/cctbx/modules/cctbx_project/scitbx/math/tests/tst_curve_fitting.py"
OK

real	0m3.611s
user	0m3.480s
sys	0m0.125s
```

if one test run alone can vary in run-time from 4-41s I do not think we can rely on the time for benchmarking these changes? Even running 10 times...

bkpoon commented:
My mistake about `fable`. I saw the "." in the command and thought the tool was applied to every directory.

As for the addition of `list` to `range`, the point is that the automated tool just changes them all regardless of if they are necessary. The unnecessary ones will usually be inside a function that accepts something iterable. In the example I provided, the list is inside a `zip` function. And as you have found, they are not only just in tests. I'm not claiming this change is causing the change in performance. I'm saying that it is not necessary to construct a list in those situations.

I have some additional times in seconds for tests on 1 core over 10 trials.

| module | old (mean/std dev) | new (mean/std dev) |
|-|-|-|
| `scitbx` | 113.7 (1.6) | 176.9 (22) |
| `cctbx` | 343.8 (1.8) | 378.3 (2.2) |

The `scitbx` tests are probably not too important since they are more tests of the C++ code than Python code. But it's not clear why the spread is larger. The server was mostly idle when the tests were run and everything is stored on a RAM disk. As for the `cctbx` tests, the times are pretty consistent and there is about a 10% difference. I'm bringing up the point that there will likely be a performance difference from these changes.

Others can chime in about the importance of these 2 points.

One other comment from looking at the changes, a lot of files have both `from six.moves import range` and `from builtins import range`. I'm not sure if there is an important difference between the two.

And as for the `list` and `dict` comment, I saw this when I was applying the second stage fixes. It might have been the `lib2to3.fixes.fix_dict` fix.

graeme-winter commented:
@bkpoon total of 18 lines match `grep -R 'list(range(' . | grep zip` so I can afford to manually remove list in those cases. Looking for the more general case of where this is inside something which can use an iterable interface is slightly harder.

graeme-winter commented:
@bkpoon I have

- removed duplicate range imports
- removed list casting where input to a flex constructor - in some cases they are needed for `list(range(n)) + [x]` or similar
- removed list in context of zip (manually) 

to confirm - no changes made outside the packages intended. Performance things could have been down to range from builtins - using six.moves should help here. Not sure how you thought I had touched `fable`?

In many of the cases which are left `list()` is needed, so think these changes are reasonable... next?

graeme-winter commented:
Re: `fable`

> My mistake about `fable`. I saw the "." in the command and thought the tool was applied to every directory.

Ah, OK, not derived from the diffs - from the commentary you assumed I had changed every file. 

Deliberately not touching as mentioned early in the thread...

ndevenish commented:
So yes, apparently the `python-future` implementation of `range` is more complete in it's python 3 behaviour for python 2, which means that it's slightly slower on python 2. Whereas `six` just renames `xrange`.

With the replacements of `builtins` to `six.moves` my `scitbx` tests actually run slight faster, overall (~140s vs ~150s)

ndevenish commented:
Fixed conflicts

bkpoon commented:
I'm at RapiData for the next 2 days, but some initial tests seem to show the performance is about the same as before.

graeme-winter commented:
Sounds positive - any comments from others on more work to do before we conclude the request?

graeme-winter commented:
Python3 is happy with compiling all the modules I have touched:

```
Grey-Area p3_compile :) $ for repo in boost_adaptbx cbflib_adaptbx cctbx chiltbx clipper_adaptbx cootbx cudatbx dxtbx fftw3tbx gltbx iotbx libtbx omptbx rstbx scitbx simtbx smtbx wxtbx; do python3 -m compileall -q /Users/graeme/svn/scr/cctbx_project/${repo} ; done
Grey-Area p3_compile :) $
```

=> syntax is now compatible

nksauter commented:
Graeme,

I will need to go through this line by line before approving.  This pull
request specifically touches my own code.

Nick

Nicholas K. Sauter, Ph. D.
Senior Scientist, Molecular Biophysics & Integrated Bioimaging Division
Lawrence Berkeley National Laboratory
1 Cyclotron Rd., Bldg. 33R0345
Berkeley, CA 94720
(510) 486-5713


On Wed, May 8, 2019 at 12:25 AM Graeme Winter <notifications@github.com>
wrote:

> Sounds positive - any comments from others on more work to do before we
> conclude the request?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cctbx/cctbx_project/pull/325#issuecomment-490378998>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADQ24VVDPX6UY3P3JOHG6G3PUJ56LANCNFSM4HJJ3UGQ>
> .
>

graeme-winter commented:
@nksauter there are a lot of changes and they touch a lot of people's code. I know @bkpoon has reviewed what I have done, you may find it useful to discuss with him in advance of reviewing individual changes.

Oeffner commented:
@graeme-winter @bkpoon @ndevenish @Anthchirp @nksauter 

FWIW I have set up a nightly Windows build and Phenix regression test of this branch. The log files are available from 
http://www-structmed.cimr.cam.ac.uk/Local/PhenixTests/Logs/
and are labelled with the short hash of the head of the branch, say tests413d51881-x64.txt. There are also log files for the regression tests of the master branch, say testsdev-3498-x64.txt. 
I hope this can be useful for someone judging if this pull request can be approved.

graeme-winter commented:
@Oeffner thank you! Practical consideration 

Looking through seems like the summary is pretty good:

Summary:
  Tests run                    : 763
  Failures                     : 4
  Warnings (possible failures) : 14
  Stderr output (discouraged)  : 4

Some of the errors are in dials, should look properly when I can get to a computer 

Anyhow, appreciate the contribution 👍

graeme-winter commented:
@Oeffner useful information would be how the branch differs from master in terms of tests successfully passing... possible? Guess I could write a script that compares later anyway...

Oeffner commented:
@graeme-winter
I guess that is possible. Maybe easiest to extract the failing ones first from both log files to see which ones cannot be paired up. 
FYI the script for the builds starts at 6am and should be done with the tests before noon I think.
Rob
From: Graeme Winter 
Sent: Wednesday, May 8, 2019 6:16 PM
To: cctbx/cctbx_project 
Cc: Robert Oeffner ; Mention 
Subject: Re: [cctbx/cctbx_project] Python 3 syntax compatibility (#325)

@Oeffner useful information would be how the branch differs from master in terms of tests successfully passing... possible? Guess I could write a script that compares later anyway...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


     Virus-free. www.avg.com

Anthchirp commented:
These kinds of things would be much easier with the xml output from run_tests_parallel. Do you have those available?

Oeffner commented:
If you can point me to what the name of the file would be I'll try to 
recover it. Or is there a flag to set for running the tests for 
generating xml output . There's a large bunch of files in the 
tests\cctbx_regression.test_nightly folder many with unconventional 
filename suffixes so I'm not sure which one of those would contain xml 
output.

Rob


On 09/05/2019 00:26, Markus Gerstel wrote:
>
> These kinds of things would be much easier with the xml output from 
> run_tests_parallel. Do you have those available?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/cctbx/cctbx_project/pull/325#issuecomment-490687270>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AD7GBUWZ4H5HWTWDRLV5FOLPUNOS5ANCNFSM4HJJ3UGQ>.
>
-- 
Robert Oeffner, Ph.D.
Research Associate, The Read Group
Department of Haematology,
Cambridge Institute for Medical Research
University of Cambridge
Cambridge Biomedical Campus
The Keith Peters Building
Hills Road
Cambridge CB2 0XY

www.cimr.cam.ac.uk/investigators/read/index.html
tel: +44(0)1223 763234

Anthchirp commented:
The file is called `output.xml` and is generated in the test folder by `libtbx.run_tests_parallel` [if the python module `junit_xml` is available](https://github.com/cctbx/cctbx_project/blob/7c7b26d5d860a2f26c8069760375b246943f3757/libtbx/test_utils/parallel.py#L160), so if you don't have that in your environment you may need a [`pip install junit-xml`](https://pypi.org/project/junit-xml/) beforehand.

bkpoon commented:
In conversation with @nksauter and @graeme-winter, we will be working on verifying the changes from the automated tool.

Oeffner commented:
Does this mean there's no longer a need for doing the nightly regression tests of this branch on Windows?

Now that I have the scripts set up and if people want I can easily select a different branch for nightly regression tests on Windows.

bkpoon commented:
We're working on verifying the automated changes and getting things running on Python 3. There will be a new pull request for that, so we can switch over when it's ready.

graeme-winter commented:
Yes the work is being done again and slightly differently from Berkeley 

As far as I’m aware there is no public branch for this so will appear as a pull request when ready or next Wednesday for work in progress?

Should be able to switch your tests once that’s done
